File: day04/day04.py
# Day 04: Scratchcards

import sys

# ==============================================================================

# If no specific input given, default to "day<X>.in"
if len(sys.argv) == 1:
  input_fname = sys.argv[0].replace(".py", ".in")
else:
  input_fname = sys.argv[1]

# Read and parse input
cards = []
with open(input_fname) as f:
  for line in f:
    parts = line.split(":")[1].split(" | ")
    winning = [int(x) for x in parts[0].strip().split(" ") if x != ""]
    own = [int(x) for x in parts[1].strip().split(" ") if x != ""]
    cards.append((winning, own))

# ------------------------------------------------------------------------------
# Part 1

cards_matches = []

tot_points = 0
for card in cards:
  winning, own = card
  matches = 0
  for num in own:
    if num in winning:
      matches += 1
  points = 2**(matches-1) if matches >= 1 else 0
  cards_matches.append(matches)
  tot_points += points

print("Part 1:", tot_points)

# ------------------------------------------------------------------------------
# Part 2

# Too slow!
# to_check = [i+1 for i in range(len(cards))]
# retired = []
# while len(to_check) > 0:
#   num = to_check.pop(0)
#   retired.append(num)
#   matches = cards_matches[num-1]
#   for n in list(range(num+1, num+matches+1)):
#     to_check.append(n)

spawns = {}
def count_spawns(num):
  if num in spawns:       # Memoization, baby!
    return spawns[num]
  matches = cards_matches[num-1]
  count = 1
  for n in range(num+1, num+1+matches):
    count += count_spawns(n)
  spawns[num] = count
  return count

tot_cards = 0
for num in range(1, len(cards)+1):
  tot_cards += count_spawns(num)

print("Part 2:", tot_cards)

File: day04/day04.jl
# Day 04: Scratchcards

# ==============================================================================

# Get input filename as CLI arg. If not given, default to "day<X>.in"
if length(ARGS) > 0
  fname = ARGS[1]
else
  fname = replace(PROGRAM_FILE, ".jl" => ".in")
end

# Parse input
cards = []
lines = readlines(fname)
 for (card_num, line) in enumerate(lines)
  winning, haves = split(split(line, ":")[2], "|")
  winning = map(x -> parse(Int, x), split(strip(winning), " ", keepempty=false))
  haves = map(x -> parse(Int, x), split(strip(haves), " ", keepempty=false))
  push!(cards, (card_num, winning, haves))
end


# ------------------------------------------------------------------------------
# Part 1

cards_matches = Dict()

tot_points = 0
for (card_num, winning, haves) in cards
  matches = 0
  for num in haves
    if num in winning
      matches += 1
    end
  end
  cards_matches[card_num] = matches
  if matches > 0
    global tot_points += 2^(matches-1)
  end
end

println("Part 1: ", tot_points)

# ------------------------------------------------------------------------------
# Part 2

cache = Dict()

function count_spawned(card_num)
  if haskey(cache, card_num)
    return cache[card_num]
  end
  count = 1
  matches = cards_matches[card_num]
  for n in card_num+1:card_num+matches
    count += count_spawned(n)
  end
  cache[card_num] = count
  return count
end 

total_cards = 0
for n in 1:length(cards)
  global total_cards += count_spawned(n)
end

println("Part 2: ", total_cards)